### PR TITLE
Read localization from QuakeEX.kpf

### PR DESCRIPTION
**Pull request by sezero**
uses a customized and highly stripped down miniz version

### CODE REVIEW

**Comment by sezero** [comment]
As an extension, if this patch is OK, we can use SDL_RWops so as not to clobber
the 64 bit file sizes/offsets miniz is using. E.g.:
```diff
diff --git a/Quake/common.c b/Quake/common.c
index 718b0f9..a006fa4 100644
--- a/Quake/common.c
+++ b/Quake/common.c
@@ -2507,9 +2507,9 @@ unsigned COM_HashString (const char *str)
 
 static size_t mz_zip_file_read_func(void *opaque, mz_uint64 ofs, void *buf, size_t n)
 {
-	if (fseek((FILE *)opaque, (long)ofs, SEEK_SET))
+	if (SDL_RWseek((SDL_RWops*)opaque, (Sint64)ofs, RW_SEEK_SET) < 0)
 		return 0;
-	return fread(buf, 1, n, (FILE *)opaque);
+	return SDL_RWread((SDL_RWops*)opaque, buf, 1, n);
 }
 
 /*
@@ -2524,6 +2524,7 @@ void LOC_LoadFile (const char *file)
 	int i,lineno;
 	char *cursor;
 
+	SDL_RWops *rw = NULL;
 	mz_zip_archive archive;
 	size_t size = 0;
 
@@ -2546,22 +2547,21 @@ void LOC_LoadFile (const char *file)
 	fp = fopen(path, "rb");
 	if (!fp)
 	{
+		Sint64 sz;
 		q_snprintf(path, sizeof(path), "%s/QuakeEX.kpf", com_basedir);
-		fp = fopen(path, "rb");
-		if (!fp) goto fail;
-		fseek(fp, 0, SEEK_END);
-		i = ftell(fp);
-		fseek(fp, 0, SEEK_SET);
-		if (i <= 0) goto fail;
+		rw = SDL_RWFromFile(path, "rb");
+		if (!rw) goto fail;
+		sz = SDL_RWsize(rw);
+		if (sz <= 0) goto fail;
 		archive.m_pRead = mz_zip_file_read_func;
-		archive.m_pIO_opaque = fp;
-		if (!mz_zip_reader_init(&archive, i, 0)) goto fail;
+		archive.m_pIO_opaque = rw;
+		if (!mz_zip_reader_init(&archive, sz, 0)) goto fail;
 		localization.text = (char *) mz_zip_reader_extract_file_to_heap(&archive, file, &size, 0);
 		if (!localization.text) goto fail;
+		mz_zip_reader_end(&archive);
+		SDL_FreeRW(rw);
 		localization.text = (char *) realloc(localization.text, size+1);
 		localization.text[size] = 0;
-		fclose(fp);
-		mz_zip_reader_end(&archive);
 	}
 	else
 	{
@@ -2573,6 +2573,7 @@ void LOC_LoadFile (const char *file)
 		{
 fail:			if (fp) fclose(fp);
 			mz_zip_reader_end(&archive);
+			if (rw) SDL_RWclose(rw);
 			Con_Printf("Couldn't load '%s'\nfrom '%s'\n", file, com_basedir);
 			return;
 		}
```

**Comment by sezero** [comment]
I went ahead and used SDL_RWops for file reads for this.

File: Quake/common.c [hunk 1]
```diff
@@ -2879,6 +2881,13 @@ unsigned COM_HashString (const char *str)
 	return hash;
 }
 
+static size_t mz_zip_file_read_func(void *opaque, mz_uint64 ofs, void *buf, size_t n)
+{
+	if (SDL_RWseek((SDL_RWops*)opaque, (Sint64)ofs, RW_SEEK_SET) < 0)
+		return 0;
+	return SDL_RWread((SDL_RWops*)opaque, buf, 1, n);
```
Comment (Novum) [inline Quake/common.c:2888]: Code is a bit inconsistent with other file IO. I've seen either just fopen or Sys_FileOpenRead

Comment (sezero) [reply Quake/common.c:2888]: SDL2 RWops works with 64 bit file sizes and offsets, so does miniz:
I didn't want to clobber miniz 64 bit values and wanted something
compatible with it.

**Comment by Novum** [comment]
Looks good. Feel free to merge (probably best to squash)

**Comment by sezero** [comment]
Done.

We probably should remove the 'extract localization file' from readme?

**Comment by ericwa** [comment]
Bit late - but this approach looks good to me 👍

**Comment by Novum** [comment]
> Done.
> 
> We probably should remove the 'extract localization file' from readme?

Done

**Pull request merged**